### PR TITLE
PLAT-129518: Sync Item with Sandstone

### PR DESCRIPTION
**Pull request by adrian-cocoara-lgp**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [x] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
Refactor Item in Agate to follow the architecture model from Sandstone:
- refactored render method for ItemContent
- refactored render method for ItemBase
- added `centered` prop for Item
- added `size` prop for Item (default large)
- updated storybook sampler
- added unit tests
- updated UI/screenshot tests

### Links
[//]: # (Related issues, references)
PLAT-129518

### Comments

Enact-DCO-1.0-Signed-off-by: Adrian Cocoara adrian.cocoara@lgepartner.com

### CODE REVIEW

File: styles/variables-base.less [hunk 1]
```diff
@@ -376,6 +377,14 @@
 @agate-item-side-outset:              @agate-component-spacing;
 @agate-item-sub-text-size:            inherit;
 @agate-item-vertical-padding:         3px;
+@agate-item-small-font-size:		  inherit;
+@agate-item-small-height:			  54px;
```
Comment (daniel-stoian-lgp) [inline styles/variables-base.less:381]: indentation needs to be fixed, if possible

File: styles/variables-gallium.less [hunk 2]
```diff
@@ -113,11 +113,15 @@
 @agate-item-font-weight:      500;
 @agate-item-height:           90px;
 @agate-item-height-large:     120px;
-@agate-item-label-font-size:  30px;
```
Comment (daniel-stoian-lgp) [inline styles/variables-gallium.less:116]: nice catch! 👍

**Review by stanca-pop-lgp — APPROVED** [review]

**Review by daniel-stoian-lgp — APPROVED** [review]

File: samples/sampler/stories/default/Item.js [hunk 3]
```diff
@@ -4,35 +4,29 @@ import UiItem, {ItemBase as UiItemBase} from '@enact/ui/Item';
...
-
 storiesOf('Agate', module)
 	.add(
 		'Item',
 		() => {
-			const slotBeforeIcon = select('slotBefore Icon', iconList, Config);
-			const slotAfterIcon = select('slotAfter Icon', iconList, Config);
 			return (
 				<Item
+					centered={boolean('centered', Config)}
```
Comment (jeonghee27) [inline samples/sampler/stories/default/Item.js:19]: image

Center prop's behavior looks unstable.
I don't know if there were any other requirements for `center`.
If you don't have a GUI requirement or other necessity, would you leave it remove the `center` prop?
Or do you want to solve the bug individually?

File: tests/screenshot/apps/components/Item.js [hunk 4]
```diff
@@ -5,7 +5,11 @@ const ItemTests = [
 	<Item>Hello Item</Item>,
 	<Item disabled>Hello Item</Item>,
 	<Item slotBeforeIcon="circlelarge" slotAfterIcon="circlelarge">Hello Item</Item>,
```
Comment (jeonghee27) [inline tests/screenshot/apps/components/Item.js:7]: `slotBeforeIcon` `slotAfterIcon` nee
Could you please add other props as well (inline, labelPosition, and so on)?

File: tests/ui/specs/Item/Item-specs.js [hunk 5]
```diff
@@ -112,7 +112,7 @@ describe('Item', function () {
 
 				it('should focus an inline item with 5-way Left', function () {
 					item7Inline.focus();
-					Page.spotlightLeft();
+					Page.spotlightDown();
```
Comment (jeonghee27) [inline tests/ui/specs/Item/Item-specs.js:115]: 'should focus an inline item with 5-way Left'

Page.spotlightDown()

I think the previous(spotlightLeft) is right. isn't it?

File: styles/variables-base.less [hunk 6]
```diff
@@ -376,6 +377,14 @@
...
 @agate-item-sub-text-size:            inherit;
 @agate-item-vertical-padding:         3px;
+@agate-item-small-font-size:          inherit;
+@agate-item-small-height:             54px;
+@agate-item-small-line-height:        @agate-item-small-height - (@agate-item-vertical-padding * 2);
+@agate-item-small-label-font-size:    inherit;
+@agate-item-small-label-line-height:  inherit;
+@agate-item-small-haslabel-padding:   12px 0;
+@agate-item-slotafter-padding-left:   inherit;
+@agate-item-small-itemcontent-line-height: inherit;
```
Comment (jeonghee27) [inline styles/variables-base.less:387]: please change order 

```
@agate-item-slotafter-padding-left:   inherit;
@agate-item-small-itemcontent-line-height: inherit;
```

File: Item/Item.js [hunk 7]
```diff
@@ -85,30 +133,159 @@ const ItemContent = kind({
...
+		 */
 		css: PropTypes.object,
+
+		/**
+		 * Applies a disabled style and the control becomes non-interactive.
+		 *
+		 * @type {Boolean}
+		 * @public
+		 */
+		disabled:PropTypes.bool,
```
Comment (jeonghee27) [inline Item/Item.js:182]:
```suggestion
		disabled: PropTypes.bool,
```

**Review by jeonghee27 — APPROVED** [review]
LGTM

File: Item/Item.js [hunk 8]
```diff
@@ -10,37 +10,92 @@
...
 
+import componentCss from './Item.module.less';
+
+/**
+ * Renders the Item content.
+ *
+ * @class ItemContent
+ * @memberof agate/Item
+ * @ui
+ * @public
```
Comment (MikyungKim) [inline Item/Item.js:35]: I think it's private.

File: Item/Item.js [hunk 9]
```diff
@@ -10,37 +10,92 @@
...
+ */
 const ItemContent = kind({
 	name: 'ItemContent',
-	propTypes: {
+
+	propTypes:  /** @lends agate/Item.ItemContent.prototype */ {
+		/**
+		 * Text displayed when passed as children.
+		 *
+		 * @type {String | Number}
```
Comment (MikyungKim) [inline Item/Item.js:44]:
```suggestion
		 * @type {*}
```

File: Item/Item.js [hunk 10]
```diff
@@ -10,37 +10,92 @@
...
 		content: PropTypes.any,
+
+		/**
+		 * Customizes the component by mapping the supplied collection of CSS class names to the
+		 * corresponding internal elements and states of this component.
+		 *
+		 * The following classes are supported:
+		 *
+		 * `itemContent` - The root class name
+		 * `content` - The content class name
```
Comment (MikyungKim) [inline Item/Item.js:56]: I don't see these in `publicClassNames`. Are you going to make it public as well?

File: Item/Item.js [hunk 11]
```diff
@@ -10,37 +10,92 @@
...
+		 *
+		 * @type {Object}
+		 * @public
+		 */
 		css: PropTypes.object,
+
+		/**
+		 * Text displayed when passed in `label` prop.
+		 *
+		 * @type {String | Number}
```
Comment (MikyungKim) [inline Item/Item.js:67]:
```suggestion
		 * @type {*}
```

File: Item/Item.js [hunk 12]
```diff
@@ -10,37 +10,92 @@
...
+		 * @public
+		 */
 		label: PropTypes.any,
-		labelPosition: PropTypes.any
+
+		/**
+		 * Repositioning of label.
```
Comment (MikyungKim) [inline Item/Item.js:73]:
```suggestion
		 * The position of the label relative to the primary content, `content`.
```

File: Item/Item.js [hunk 13]
```diff
@@ -10,37 +10,92 @@
...
+		 *
+		 * @type {String}
```
Comment (MikyungKim) [inline Item/Item.js:75]:
```suggestion
		 * @type {('above'|'after'|'before'|'below')}
```

File: Item/Item.js [hunk 14]
```diff
@@ -85,30 +141,159 @@ const ItemContent = kind({
...
+		 * @private
+		 */
 		selected: PropTypes.bool,
+
+		/**
+		 * The size of the item.
+		 *
+		 * @type {('large'|'small')}
+		 * @default 'large'
+		 * @private
```
Comment (MikyungKim) [inline Item/Item.js:237]: public?

File: Item/Item.js [hunk 15]
```diff
@@ -85,30 +141,159 @@ const ItemContent = kind({
...
+		 *
+		 * @type {Node}
+		 * @public
+		 */
 		label: PropTypes.node,
+
+		/**
+		 * The position of the label relative to the primary content, `children`.
+		 *
+		 * @type {('above'|'after'|'before'|'below')}
```
Comment (MikyungKim) [inline Item/Item.js:211]:
```suggestion
		 * @type {('above'|'after'|'before'|'below')}
		 * @default 'below'
```

File: Item/Item.js [hunk 16]
```diff
@@ -85,30 +141,159 @@ const ItemContent = kind({
...
+
 	computed: {
-		className: ({selected, styler}) => styler.append({selected})
+		className: ({selected, centered, size, styler}) => styler.append(
+			{
+				selected,
+				centered
+			},
+			size
+		),
```
Comment (MikyungKim) [inline Item/Item.js:283]: I think some of sandstone/Item structure didn't sync in this PR.
`hasLabel` should be placed here, not in `ItemContent`.
Item.modules.less should have the proper structure along with this change.
Would you take a look, please?

File: Item/Item.js [hunk 17]
```diff
@@ -133,36 +320,40 @@ const ItemBase = kind({
...
-	Slottable({slots: ['label', 'slotAfter', 'slotBefore']}),
-	Pure,
+	MarqueeController({marqueeOnFocus: true, invalidateProps: ['inline']}),
 	UiItemDecorator,
-	Spottable,
-	MarqueeController({marqueeOnFocus: true}),
-	Skinnable
+	Slottable({slots: ['label', 'slotAfter', 'slotBefore']}),
+	Skinnable,
+	Spottable
```
Comment (MikyungKim) [inline Item/Item.js:338]: The order is different from sandstone's. Is there a particular reason for this?

File: tests/screenshot/apps/components/Item.js [hunk 18]
```diff
@@ -1,11 +1,24 @@
...
+
+const slotBeforeIcon = <Icon>circlelarge</Icon>;
+const slotAfterIcon = <Icon>circlelarge</Icon>;
 
 const ItemTests = [
 	<Item>Hello Item</Item>,
 	<Item disabled>Hello Item</Item>,
-	<Item slotBeforeIcon="circlelarge" slotAfterIcon="circlelarge">Hello Item</Item>,
-	<Item label="label text" labelPosition="below">Hello Item</Item>
+	<Item slotBefore={slotBeforeIcon}>Hello Item</Item>,
```
Comment (MikyungKim) [inline tests/screenshot/apps/components/Item.js:11]: I hope we could add more tests here.

**Review by MikyungKim — CHANGES_REQUESTED** [review]
I'm sorry for the delayed review.
Please find my comments and verify the ui tests, ss tests, and docs result too.
It's quite a big change so I think we need to be careful here. :)

File: Item/Item.js (hunk 10 above)
Comment (adrian-cocoara-lgp) [reply Item/Item.js:56]: No, I won't. I thought app developers should know which classNames contains `ItemContent`. I will remove these two and only leave `label`. Do you think this is right?

File: Item/Item.js (hunk 16 above)
Comment (adrian-cocoara-lgp) [reply Item/Item.js:283]: You're right, I missed this one. `hasLabel` is placed in `ItemBase` now. Thanks for noticing!

File: Item/Item.js (hunk 17 above)
Comment (adrian-cocoara-lgp) [reply Item/Item.js:338]: The reason was to order these alphabetically. I can change it back to match Sandstone.

File: tests/screenshot/apps/components/Item.js (hunk 18 above)
Comment (adrian-cocoara-lgp) [reply tests/screenshot/apps/components/Item.js:11]: More tests have been added 😅

File: Item/Item.js (hunk 10 above)
Comment (MikyungKim) [reply Item/Item.js]: Yeap. We only list the classes which are listed in `publicClassNames` here.

**Comment by MikyungKim** [comment]
@adrian-cocoara-lgp , I fixed the lint errors and some of jsdocs. Also, I found `.itemContent` className usage is quite different from sandstone's so I revised `Item.module.less`. When I tested with `centered` prop and very long content and label, the layout seems broke so I didn't apply additional width change as sandstone did. But applying width change is more natural when `orientation` is `horizontal` so I left it as is without `shrink` prop like the original code.
UI tests were passed locally and I'm testing screenshot tests on Jenkins. If the result looks good, I'll merge this.

**Review by MikyungKim — APPROVED** [review]
:shipit:

**Pull request merged**